velox/expression/SwitchExpr.cpp

```diff
@@ -140,27 +141,33 @@ void SwitchExpr::evalSpecialForm(
   // Evaluate the "else" clause.
   if (remainingRows.get()->hasSelections()) {
     if (hasElseClause_) {
-      inputs_.back()->eval(*remainingRows.get(), context, result);
+      inputs_.back()->eval(*remainingRows.get(), context, localResult);
     } else {
-      context.ensureWritable(*remainingRows.get(), type(), result);
+      context.ensureWritable(*remainingRows.get(), type(), localResult);
 
       // fill in nulls for remainingRows
       remainingRows.get()->applyToSelected(
-          [&](auto row) { result->setNull(row, true); });
+          [&](auto row) { localResult->setNull(row, true); });
     }
   }
-  // Some rows may have not been evaluated by any then or else clause because a
-  // condition threw an error on these rows. We make sure the result vector has
-  // at least the size of rows.end().
-  if (context.errors() && (!result || result->size() < rows.end())) {
-    if (result && result->isConstantEncoding() && result.unique()) {
-      result->resize(rows.end());
-    } else {
+
+  // Some rows may have not been evaluated by any then or else clause because
+  // a condition threw an error on these rows. We make sure of the following:
+  // 1. The result vector has at least the size of rows.end().
+  // 2. The result should be addressable for the rows where an error occurred.
+  // If the type is not fixed primitive we do this by adding nulls or extending
+  // constant vector size.
+
+  bool isPrimtiveFied = type()->isPrimitiveType() && type()->isFixedWidth();
+  if (!localResult || localResult->size() < rows.end() || !isPrimtiveFied) {
```

Comment:
do we need this check? we should do this addition of nulls regardless and check for if(context.errors() || decoded.mayhavenulls()) => add both nulls and error rows here.

Even for primitive fields with fixed width, we can have a dict layer over it where the nulled out or error rows can have invalid indices

Reply:
addressed